### ISSUE TITLE
Union subtraction limitation

### ISSUE THREAD

**Issue by krzkaczor** [astral-sh/ty#4442]
### Summary

It's impossible to make ty subtract generic union members like this:

[Playground](https://play.ty.dev/f4c3dba6-c8cb-46a9-b8c4-1bf9798f4603)
```py
from typing import Never, assert_type


class Box[R]:
    def get(self) -> R:
        raise NotImplementedError


box: Box[int | str] = Box()
one = int(1)

def provide_oneshot[T, R2 = Never](box: Box[T | R2], impl: T) -> Box[R2]:
    raise NotImplementedError


# EXPECTED: accepted, good1 is Box[str] (T = int, R2 = str).
# ACTUAL:   invalid-argument-type "Expected Box[int], found Box[int | str]"
#           — R2 is never solved from the union match and defaults to Never —
#           and the assert_type reports the inferred type as Box[Never].
good1 = provide_oneshot(box, one)
assert_type(good1, Box[str])
```

### Version

ty 0.0.75 (a553dcc1a 2026-08-26)

**Comment by carljm** [comment]
The `invalid-argument-type` here is a bug; it should be fixed by the union half of #3557.

That said, I'm not totally sure if we would end up inferring `[T = int, R2 = str]` as desired here, or `[T = int | str, R2 = Never]`; both are valid.

One workaround that does give you your desired result today is to use a protocol which matches `Box` rather than `Box` itself in the signature of `provide_oneshot`:

```py
from typing import Never, Protocol, assert_type

class Box[R]:
    def get(self) -> R:
        raise NotImplementedError

class Source[R](Protocol):
    def get(self) -> R: ...

def provide_oneshot[T, R2 = Never](
    box: Source[T | R2], impl: T
) -> Box[R2]:
    raise NotImplementedError

box: Box[int | str] = Box()
one = int(1)

result = provide_oneshot(box, one)
assert_type(result, Box[str])  # Passes on current main
```

In fact ty is the only type checker that gives `Box[str]` for this version, which does suggest that once we fix this for the nominal-only version, we may still get the same result.
